### PR TITLE
demo: Client demo for Kotlin application using Ktor.

### PR DESCRIPTION
**Pull request by ybelMekk**
Demo for clients (ktor n kotlin) using token-support. Add support demo for overriding `JwtBearerTokenResolver` when using an Identity provider requires a jwt with different claims than standard. Also added demo on `OAuth2HttpClient` to override the form-post .

### CODE REVIEW

**Comment by ybelMekk** [comment]
Can merge this one together / or separate for release draft with all dependencies that are pending for the project?

File: README.md [hunk 1]
```diff
@@ -239,6 +239,114 @@ Add the modules that you need as Maven dependencies.
 - **`no.nav.security.jwt.issuer.[issuer shortname].cookiename`** - The value of the cookie containing a ID token (not required, only necessary if your api receives calls directly from a browser)
 - **`no.nav.security.jwt.issuer.[issuer shortname].validation.optional_claims`** - A comma separated list of optional claims that will be excluded from default claims.
 
+### token-client-core
+Module provides a core token client support with supported grants: `jwt_bearer`, `client_credentials`, `token_exchange`, through which applications can gain an access token.
```
Comment (jksolbakken) [inline README.md:243]: Funky use of commas. Something like "The module provides a core token client support with supported grants: `jwt_bearer`, `client_credentials` and `token_exchange` through which applications can gain an access token." ?

File: README.md [hunk 2]
```diff
@@ -239,6 +239,114 @@ Add the modules that you need as Maven dependencies.
 - **`no.nav.security.jwt.issuer.[issuer shortname].cookiename`** - The value of the cookie containing a ID token (not required, only necessary if your api receives calls directly from a browser)
 - **`no.nav.security.jwt.issuer.[issuer shortname].validation.optional_claims`** - A comma separated list of optional claims that will be excluded from default claims.
 
+### token-client-core
+Module provides a core token client support with supported grants: `jwt_bearer`, `client_credentials`, `token_exchange`, through which applications can gain an access token. 
+The module generates a client_assertion based on application configuration properties in: `ClientProperties`. A ClientAssertion is signed with configured `client_jwk` can be used or a `client_secret`.
+Use the interface `JwtBearerTokenResolver` to override JWT used in specified grant request. By extending the `OAuth2HttpClient`, you can override the form-post to a provider, specifying headers and or form-parameters.
```
Comment (jksolbakken) [inline README.md:245]: Superflous comma after `OAuth2HttpClient`

File: token-client-kotlin-demo/src/main/kotlin/no/nav/security/token/support/ktor/oauth/TokenResolver.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import no.nav.security.token.support.client.core.context.JwtBearerTokenResolver
+import no.nav.security.token.support.ktor.jwt.ClientAssertion
+import java.util.Optional
+
+class TokenResolver(
+    private val client: ClientAssertion
+) : JwtBearerTokenResolver {
+
+    // Override default client_assertion jwt, with specified Idp jwt
+    override fun token(): Optional<String> {
```
Comment (jksolbakken) [inline token-client-kotlin-demo/src/main/kotlin/no/nav/security/token/support/ktor/oauth/TokenResolver.kt:12]: Why `Optional` ?

**Review by jksolbakken — APPROVED** [review]

**Review by tronghn — APPROVED** [review]
lgtm

File: token-client-kotlin-demo/src/main/kotlin/no/nav/security/token/support/ktor/oauth/TokenResolver.kt (hunk 3 above)
Comment (tronghn) [reply token-client-kotlin-demo/src/main/kotlin/no/nav/security/token/support/ktor/oauth/TokenResolver.kt:12]: The interface implemented has a signature of `Optional<String>`

**Review by tommytroen — APPROVED** [review]

**Pull request merged**